### ISSUE TITLE
`{}` should not be assignable to `object`

### ISSUE THREAD

**Issue by kirkwaiblinger** [microsoft/TypeScript#60582]
### 🔎 Search Terms

empty object type, `{}`, type safety violation, unsound

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about [object types](https://www.typescriptlang.org/docs/handbook/2/objects.html) and [structural typing](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#structural-typing)

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.8.0-dev.20241124#code/GYVwdgxgLglg9mABFAhgawKYGcDyAjAKw2gAoAPALkTkOKgEpEBvAKEXcQHpPEAVACxhZEAJxgBzflAA2AT0RDEKPgGVEGESLgiAdGw4xgicogC85xAAZGrDncQQEWONIw6NWkSQDkAxXhEUSH5ELH44EGkAE0Q8DERwEQwUCH4UPFdvegBufXYAXxZClkcwLChqKiZ8sytclm5EACEQCqhBYSEwbygAGiUwGLCI6Ni3FlRMXFpSOBygA

### 💻 Code

```ts
function takesObject(x: object) {
    // This rightly is a TS error.
    if (x === 0) {
        console.error('This branch should be unreachable');
    }
}

const o: {} = 0;

// But this isn't, and should be.
takesObject(o);
```

### 🙁 Actual behavior

No error on `takesObject(o)`. `{}` can be assigned to type `object`, despite `{}` meaning "all nonnullish values", whereas `object` means only JS object types.

### 🙂 Expected behavior

Error on `takesObject(o)`. `{}` is a wider type than `object`.

### Additional information about the issue

This stems from a typescript-eslint investigation into making no-unnecessary-condition be more correct around possibly-falsy "object types", such as `{}`, or even `{ toFixed(): string}` (to which `number` may be assigned). See https://github.com/typescript-eslint/typescript-eslint/pull/10378. This also relates to previous (controversial) conversations about whether to flag the `{}` type with the linter, see, e.g. https://github.com/typescript-eslint/typescript-eslint/issues/8700. 

I'm wondering if this was simply an oversight in https://github.com/microsoft/TypeScript/pull/49119, which aimed to fix soundness holes with `{}`?

**Comment by jcalz** [comment]
duplicate #56205

**Comment by JoshuaKGoldberg** [comment]
Even if it's a duplicate, given https://github.com/microsoft/TypeScript/issues/56205#issuecomment-1777532315: 

> Allowing `{}` to be used as `object` is an intentional compatibility hole since `object` didn't always exist, so people used `{}` as the next-best thing.

I read this new issue as a request to _not_ do that. As in, now that TypeScript has had better handling around `{}` and `object` for a while (refs: https://github.com/microsoft/TypeScript/pull/49119, https://github.com/microsoft/TypeScript/issues/57735#issuecomment-2000069787). Over in typescript-eslint we've overhauled our lint rules around the empty object types to respect the new behaviors and best practices (https://github.com/typescript-eslint/typescript-eslint/issues/8700 -> https://typescript-eslint.io/rules/no-empty-object-type). This hole is a sticking point.

**Comment by jcalz** [comment]
Is it just `{}`? So would `{a: string}` still be assignable to `object` or not? What about `{length: number}`?

**Comment by JoshuaKGoldberg** [comment]
Personally, I'm of a "strict" interpretation: that any `{ ... }` type coincidentally assignable to a non-`object` primitive should not be assignable to `object`. Only with `object &` should they be. Quick examples:

* `{}`: as discussed
* `{ length: number }`: because of `string`
* `{ toPrecision(): string }` and  `{ toPrecision(precision?: number): string }`: because of `number`

This is probably tricky from an implementation perspective. But it's _right_.

**Comment by fatcerberus** [comment]
@JoshuaKGoldberg So you're suggesting that this shouldn't be allowed:

```ts
let obj: { length: number } = { length: 10 };
const o: object = obj;
```

Simply because `obj` could also hypothetically be a `string`?

**Comment by jakenvac** [comment]
Maybe I don't understand the nuances of `{}` but it's essentially `any` with the restriction that it does not allow nullish values, right?

If that assumption is correct, then why shouldn't `{}` be assignable to `object`? `any` is assignable to `object` because we've told the type checker that we know better than it in this scenario. Should it not the same case for `{}`?

To recontextualize the first example:
```ts
function takesObject(x: object) {
    // This rightly is a TS error.
    if (x === 0) {
        console.error('This branch should be unreachable');
    }
}

// Using any here instead of {}
const o: any = 0;

// Would you still expect it to error here? 
takesObject(o);
```

Is this more a case of `{}` being misleading syntax?

**Comment by fatcerberus** [comment]
@kirkwaiblinger If we're unpacking this in detail, negative - your example should be allowed 🙂

Yes, that was the point of the comment - it was a counterexample for @JoshuaKGoldberg's original argument.

**Comment by kirkwaiblinger** [comment]
> Yes, that was the point of the comment - it was a counterexample for [@JoshuaKGoldberg](https://github.com/JoshuaKGoldberg)'s original argument.

...and I'm saying I don't see how it's a counterexample at all? It's just an edge case that would need to be handled by the same sleight-of-hand that TS already includes in similar situations.

Regardless, this is perhaps an implementation detail consideration that's not worth spending too many more words going back and forth on before we even find out if this issue will be summarily closed as a duplicate or not 😆. So I'll leave it there for now to avoid noise in the issue conversation. But definitely an important detail to handle if this does get to the point of implementation! 🙂

**Comment by kirkwaiblinger** [comment]
I am _enthusiastically_ on board with considering how to make it so that `{ ... }` types _are_ objects! I only framed the issue in the way I did on the assumption that that would be a nonstarter. But I agree with every word in https://github.com/microsoft/TypeScript/issues/60582#issuecomment-2502179583 🙂

I guess, before thinking _too_ deeply to reinvent the wheel, I would be pretty interested to know how far down the rabbit hole you guys have gone when pondering this previously, and what undesirable edge cases might have arisen. Do you have references on this that I might look into?

My first thought is that the overarching model here would be that the _runtime JS type_, i.e. the result of `x === null ? 'null' : typeof x)` would additionally be tracked in TS's type system. If for niche usages a `Structural<T>` intrinsic type needs to be provided, it could return the exact same type but with any associated runtime type (and therefore, for example, `Structural<BoxedNumber>` would be more or less identical to `Structural<number>`, apart from possibly readonlyness).

Since I can't help it, I've started to think through some details and consequences of this, but, as mentioned, probably it's best to catch up to where you are first.

Excited to be having this conversation!

**Comment by HansBrende** [comment]
Seems to me that anytime `interface` is declared, there should be an _implicit_ `extends object` or `& object` tacked on.

The typescript built-in primitive types should simply automatically remove that modifier from their corresponding reference interfaces internally (I don't think users themselves need the ability to remove that modifier since we can't create new primitive types! There can only ever be the built-in primitives, right?)


Without an `interface` declaration however, it seems far more elegant to not assume anything. For example: `type X = {}` should not extend `object` automatically, in my opinion: it makes the inheritance hierarchy non-transitive.

However, I would also echo @kirkwaiblinger's request for references to any discussion/pitfalls/undesirable edge cases that have already been covered on this topic: would love to read more into it.

**Comment by RyanCavanaugh** [comment]
I couldn't find any relevant links (unfortunately this isn't very searchable). The root problem is that the inconsistencies here can't be *removed*, they can just be *moved*. We say that this program is legal
```ts
function foo(s: string) {
  return s.length;
}
```
But *why* is it legal? It's legal because the `string` type also includes properties from the global `String` type.

Similarly, we allow you to write this program, for the same reason:
```ts
function foo<T extends { length: number }>(x: T) {
  return x.length;
}
// Legal call
foo("hello world");
```

The proposal here implies breaking this consistency: You can no longer take some block of expressions and talk about them in a higher-order fashion if some of those expressions involve property access on primitives. Accessing the `length` of a string is now a *fundamentally different* operation than accessing the `length` of an array.

This raises further problems, consider something like this
```ts
const p = "foo";
type X = (typeof p)["length"];

type GetLength<T> = T extends { length: infer U } : U ? never;
type Y = GetLength<typeof p>;
``
Is the type `X` even legal? One argument says yes, because it's what you'd get if you wrote `p.length`. Another argument says no, because this should be basically equivalent to `GetLength`, which would now produce `never` instead.

Every place we produce or ingest a property name, you now have to think about whether that property name exists as part of an object syntax or doesn't. e.g. what's `keyof string`?

This also implicitly breaks common "branding" patterns like `string & { isNormalized: true }` since `string & object` is obviously `never`.

I'd also just weigh in that making `interface` implicitly extend `object` but not `type X = { }` seems extremely inconsistent and, arguably, counterintuitive? If anything I'd choose the opposite. You can imagine making it legal to write `interface Foo extends object {` (and making that idiomatic in your codebase) but there's not as much syntactic clarity in having to jam in an `& object` into a `type` declaration. But either way it introduces "yet another thing you have to know", which isn't great from the perspective of keeping the language learnable.

**Comment by RyanCavanaugh** [comment]
Other issues that became apparent when making an extremely rough prototype (baseline results [here](https://github.com/microsoft/TypeScript/commit/20359b9e4c987ed7fa41229ca88d715720347619))

Today `unknown` can be freely decomposed to and from the union `{} | null | undefined`. This would have to instead be the union `{} | null | undefined | string | boolean | number | symbol | bigint`. Not a dealbreaker but it's another complication to deal with.

Many JSDoc comments use `Number` instead of `number`, so that would be breaking almost everywhere. We could reinterpret `Number` in that context (we already do for other types) but it's another inconsistency.

`string` would no longer be `Iterable<>`; arguably this is a good thing because iterating a string is almost never a good idea, but if you really did mean to write a `for/of` loop on some programmatic data, it's unclear what you would do instead.

I think it'd be super weird to have `string` not be `{ [index: number]: string }`, but that's a straightforward consequence.

The definition of `NonNullable` would have to change. Unclear what the downstream implications of that would be or if that would be a dealbreaker since we can't change the lib types based on compiler settings. I'd have to think about it more.

There's some *very* real inconsistency around inference of destructuring parameters. Let's say you write
```ts
function foo({ length }) { }
foo("bar");
```
This is legal; if you look at the inferred type of `foo` it's
```ts
function foo(arg: { length: any }) : void
```
so the call would become illegal. IMO that's pretty weird, since if you had written the equivalent downleveling, you'd be fine. Maybe we could iterate through all the primitives and add them to the signature if they'd be compatible *and* overlappy with the parameter type, but that seems like spooky action at a distance.

**Comment by HansBrende** [comment]
@RyanCavanaugh unless I'm mistaken, all of the polymorphic inconsistency issues as well as the branding pattern/opaque type-alias issues you mentioned only exist in the world where `{} == object`. Hence why I suggested that **not** be the case, but rather the other way around: `{} & object` when a reference type is actually required (in most cases though, it is probably _not required_... the only place in (for example) the standard library I can even think of where `object` is a required input is `Object.create` and `Object.setPrototypeOf`!)

So we come to the problem where the vast majority of functions _do not need to assume_ `object` in their parameter types, while the vast majority of custom datatypes are in fact inhabited by `object` instances.

I suggested `interface` implicitly extend object simply to make life easier in the latter regard (it may be anecdotal, but usually when I see interface declarations, they are built with a non-primitive type in mind). However, if that would pose too much of an inconsistency, then allowing `interface extends object` syntax would **absolutely** be a viable alternative in my mind.

You also noted that `{ ... } & object` wouldn't have much syntactic clarity. That sentiment is confusing to me since the intersection operator seems to be so idiomatic. We have `{} & T` to narrow to defined types, so I'm not sure why narrowing to reference types wouldn't be in keeping with the existing landscape.  However, one could certainly imagine an alternative syntax such as: 
```typescript
type X = object { ... }
```
 if that would make things clearer! Again though, the number of places where a type signature will actually _require_ `object` as an input to a function I would think is minimal (I'm anecdotally looking at the standard lib). So I don't think people would need to care about tacking on `object` very often. (But correct me if I'm wrong.)

**Comment by HansBrende** [comment]
Another possibility I'd like to throw out to reduce possible boilerplate issues (keeping in mind, as stated above, this is not necessary for the proposal to work--_I'm just targeting the only pain-point mentioned_): `interface extends object` is valid syntax, and `interface extends object` is the _default_ extends clause, but only if there is no explicit `extends` clause. Users can opt out of this by explicitly stating `interface extends unknown`, or `interface extends XYZ` to inherit the behavior of interface `XYZ`.

**Comment by kirkwaiblinger** [comment]
@RyanCavanaugh I really appreciate the info you've provided. This is still on my mind, and I'm hoping to be able to give it a deep think when I have time, and also take a closer look at your prototype. Hoping you don't mind if this issue stays open for a bit 🙏

**Comment by I-Enderlord-I** [comment]
I find myself mostly agreeing with @HansBrende on this issue. However I do not think extending `object` should be the default in any *type* context.

As mentioned before there are too many existing uses of types in contravariant position where the restriction to `object` is not desired. This includes parameter types marked with interfaces. Having to manually opt out of an implicit `extends object` seems like a rather breaking change.

Besides that TS mostly considers types not as representations of the actual runtime structure, but as restrictions on it. From a more practical perspective, this means that types tell you what you can do with an object, not what it is. So the type `{}` and it's corresponding interface just tell you that there is no property you know how to interact with. Thus, it makes sense that these types would be something that can have properties (non-nullish), but not necessarily extend `object`. The same goes for
```ts
interface X {
  length: number;
}
```
It just describes the ability to read and store a `number` in the `length` property, not that the containing object is non-primitive. Actually needing an non-primitive object is rare and mostly comes down to calling internal library functions with that requirement.

As such I think the ideal sollution is to keep all interfaces and object types possibly primitive and only add that requirement when `& object` or the proposed `extends object` is added. After all, being non-primitive is almost never (correctly) used in practice. @HansBrende already mentioned `Object.create` and `Object.setPrototypeOf` and I agree that I have barely seen other use cases for it.

To cover the problem of obviously non-primitive objects not being assignable to `object`, having (JS) object literals, classes and all things with call and construt signatures (you all seem to forget that functions also extend `object`) automatically extend `object` implicitly would not break anything and allow passing them to functions specifically demanding `object`. In other cases, one *should not* expect an actual object anyways and use type guards like `typeof x === 'object'` to be safe, which already does narrow to `object | null`, the most common appearance of the `object` type as far as I have seen.

In the rare cases one actually demands an argument or property to be/stay strictly non-primitive, it is always possible to add the `& object` or `extends object` explicitly. As the rare exception, this should be the explicit case.

To summarize, this would mean leaving almost everything the same except:
- Have all object literals implicitly intersect `object` when checked as parameter, written onto a variables, set as property in another literal or used as expression in some other way (e.g. `({ a(this: object) {} }).a()`).
- Only consider `object`, `never`, `any` and things containing call or construt signatures assignable to `object`. All other types must intersect `object` to be used as such (this includes `{}` to come back to the motivating example of this issue).
- Ideally, introduce `extends object` for interfaces as you suggested to avoid the need for another type with `& object` appended. Possibly add this to a lot of standard library interfaces for built-in types like `Object` or `Number` and so on.
